File: app/models/child.rb
class Child < ActiveRecord::Base
#  belongs_to :club
  belongs_to :team

  validates_presence_of :first_name, :last_name

  has_many :associated_steps, inverse_of: :child
  has_many :completed_steps,
           -> { order(:rgt).uniq },
           through: :associated_steps,
           class_name: "Step",
           source: :step

  has_many :award_children
  has_many :awards, through: :award_children
  has_many :completed_awards, class_name: "Award", through: :completed_steps, source: :awards

  def pending_awards
    completed_awards.where.not(id: award_ids)
  end

  def pending_steps
    pending = Step.order(:rgt)\
                  .where.not(id: completed_step_ids)
    if completed_steps.present?
      pending = pending.where("steps.rgt > ?", completed_steps.last.rgt)
    end
    pending
  end

  def pending_step_ids
    pending_steps.ids
  end

  def steps
    all_steps = []
    all_steps << associated_steps
    all_steps << pending_steps.map do |pending|
      AssociatedStep.new child_id: self.id, step_id: pending.id
    end

  end

  def next_step
    pending_steps.first
  end

  def next_step_id
    next_step.id
  end

  def name
    "#{first_name} #{last_name}"
  end

  def team_name
    team.name if team.present?
  end

end

File: app/models/child.js
import DS from 'ember-data';

export default DS.Model.extend({
  firstName: DS.attr('string'),
  lastName: DS.attr('string'),
  nickname: DS.attr('string'),
  birthday: DS.attr('date'),
  avatarHash: DS.attr('string'),
  team: DS.belongsTo('team'),
  completedSteps: DS.hasMany('step', { inverse: 'completedBy' }),
  pendingSteps: DS.hasMany('step'),
  nextStep: DS.belongsTo('step'),
  completedAwards: DS.hasMany('award'),
  pendingAwards: DS.hasMany('award'),
  awards: DS.hasMany('award', { inverse: 'children' }),
  filteredCompletedSteps: function(){
    return this.get('completedSteps').filterBy('isLeaf', false);
  }.property('completedSteps.[]'),
  filteredPendingSteps: function(){
    return this.get('pendingSteps').filterBy('isLeaf', false);
  }.property('pendingSteps.[]'),
  avatarUrl: function() {
    var hash = this.get('avatarHash') ? this.get('avatarHash') : this.get('name');
    return "http://robohash.org/"+hash+".png?";
  }.property('avatarHash'),
  badgeAvatar: function(){
    return this.get('avatarUrl') + "&size=40x40";
  }.property('avatarUrl'),
  chipAvatar: function(){
    return this.get('avatarUrl') + "&size=32x32";
  }.property('avatarUrl'),
  largeAvatar: function(){
    return this.get('avatarUrl') + "&size=150x150&bgset=bg1";
  }.property('avatarUrl'),
  name: function() {
    return this.get('firstName') + ' ' + this.get('lastName');
  }.property('firstName', 'lastName')
});
